File: frontend/src/com/c2s.ts
import { Mapper } from '../util'
import { ProjectId, FileId } from './types'

export type Client2Server = Mapper<{
  CreateProject: {
    project_name: string
  }
  Project: {
    id: ProjectId
    msg: Client2ServerProject
  }
}>

export type Client2ServerProject = Mapper<{
  JoinProject: {}
  LeaveProject: {}
  CreateFile: {
    file_name: string
  }
  Reorder: {
    id: FileId
    new_index: number
  }
  File: {
    id: FileId
    msg: Client2ServerProjectFile
  }
}>

export type Client2ServerProjectFile = Mapper<{
  JoinFileSource: {}
  LeaveFileSource: {}
  EditFileSource: { contents: string }
  JoinFileDoc: {}
  LeaveFileDoc: {}
}>

File: src/c2s.rs
use serde::Deserialize;

use crate::file::FileId;
use crate::project::ProjectId;

#[derive(Deserialize, Debug)]
#[serde(tag = "type")]
pub enum Client2Server {
    CreateProject {
        project_name: String,
    },
    Project {
        id: ProjectId,
        msg: Client2ServerProject,
    },
}

#[derive(Deserialize, Debug)]
#[serde(tag = "type")]
pub enum Client2ServerProject {
    JoinProject,
    LeaveProject,
    CreateFile {
        file_name: String,
    },
    Reorder {
        id: FileId,
        new_index: usize,
    },
    File {
        id: FileId,
        msg: Client2ServerProjectFile,
    },
}

#[derive(Deserialize, Debug)]
#[serde(tag = "type")]
pub enum Client2ServerProjectFile {
    JoinFileSource,
    LeaveFileSource,
    EditFileSource { contents: String },
    JoinFileDoc,
    LeaveFileDoc,
}
